src/vm.rs

```diff
@@ -591,45 +600,55 @@ impl<'a, E: UserDefinedError, I: InstructionMeter> EbpfVm<'a, E, I> {
     /// // On running the program this syscall will print the content of registers r3, r4 and r5 to
     /// // standard output.
     /// let mut syscall_registry = SyscallRegistry::default();
-    /// syscall_registry.register_syscall_by_hash(6, BpfTracePrintf::call).unwrap();
+    /// syscall_registry.register_syscall_by_hash(6, BpfTracePrintf::init::<u64, UserError>, BpfTracePrintf::call).unwrap();
     /// // Instantiate an Executable and VM
     /// let config = Config::default();
     /// let mut bpf_functions = std::collections::BTreeMap::new();
     /// register_bpf_function(&config, &mut bpf_functions, &syscall_registry, 0, "entrypoint").unwrap();
     /// let mut executable = Executable::<UserError, TestInstructionMeter>::from_text_bytes(prog, None, config, syscall_registry, bpf_functions).unwrap();
     /// let mut vm = EbpfVm::<UserError, TestInstructionMeter>::new(&executable, &mut [], &mut []).unwrap();
     /// // Bind a context object instance to the previously registered syscall
-    /// vm.bind_syscall_context_object(Box::new(BpfTracePrintf {}), None);
+    /// vm.bind_syscall_context_objects(0, None);
     /// ```
-    pub fn bind_syscall_context_object(
+    pub fn bind_syscall_context_objects<C: Clone>(
         &mut self,
-        syscall_context_object: Box<dyn SyscallObject<E> + 'a>,
+        syscall_context: C,
         hash: Option<u32>,
     ) -> Result<(), EbpfError<E>> {
-        let fat_ptr: DynTraitFatPointer = unsafe { std::mem::transmute(&*syscall_context_object) };
         let syscall_registry = self.executable.get_syscall_registry();
-        let slot = match hash {
-            Some(hash) => {
-                syscall_registry
-                    .lookup_syscall(hash)
-                    .ok_or(EbpfError::SyscallNotRegistered(hash as usize))?
-                    .context_object_slot
+
+        for (_hash, syscall) in syscall_registry.entries.iter() {
+            let syscall_object_init_fn: SyscallInit<C, E> =
+                unsafe { std::mem::transmute(syscall.init) };
+            let syscall_context_object: Box<dyn SyscallObject<E> + 'a> =
+                syscall_object_init_fn(syscall_context.clone());
+            let fat_ptr: DynTraitFatPointer =
+                unsafe { std::mem::transmute(&*syscall_context_object) };
+
+            let slot = match hash {
```

Comment:
And the parameter `hash` is used instead across all entries ...

Reply:
Seems like we are lacking in test coverage for passing a non-`None` hash.  I didn't see any place in rbpf or solana that we pass a non-`None` hash, can you point me to it?

Reply:
I agree that we don't need it anymore ant thus the parameter should be removed.

But there was one case in the CLI which I think you missed: https://github.com/solana-labs/rbpf/pull/297

Reply:
I see that, are you going to remove that?  If not we need test coverage for it.

Reply:
I don't think that there is anything special about the usage in the CLI, it can be initialized like the others in example and test code.